diff --git a/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/replication/okhttp/OkHttpReplicationClient.java b/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/replication/okhttp/OkHttpReplicationClient.java
index 3281fa2bf9..36b2b9d583 100644
--- a/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/replication/okhttp/OkHttpReplicationClient.java
+++ b/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/replication/okhttp/OkHttpReplicationClient.java
@@ -291,7 +291,11 @@ public class OkHttpReplicationClient implements HttpReplicationClient {
 
 
     private boolean isUseGzip(final Map<String, String> headers) {
-        final String rawAcceptEncoding = headers.get(HttpHeaders.ACCEPT_ENCODING);
+        String rawAcceptEncoding = headers.get(HttpHeaders.ACCEPT_ENCODING);
+
+        if (rawAcceptEncoding == null) {
+            rawAcceptEncoding = headers.get(HttpHeaders.ACCEPT_ENCODING.toLowerCase());
+        }
 
         if (rawAcceptEncoding == null) {
             return false;
